gatereader/reader.py

```diff
@@ -0,0 +1,187 @@
+import serial
+import struct
+from enum import Enum
+from .utils.structparse import mystruct, t, STRUCT_FORMAT
+
+class PacketId(Enum):
+    INVALID_PACKET        = 0x00
+    ANSWER_TO_RESET       = 0x01
+    CONTINUE_BOOT         = 0x02
+    FIRMWARE_UPGRADE      = 0x03
+    FW_UPGRADE_READY      = 0x04
+    FW_WRITE              = 0x05
+    FW_WRITE_ACK          = 0x06
+    FW_UPDATE_FINISH      = 0x07
+    GET_STATUS            = 0x08
+    SET_LED               = 0x09
+    ACK                   = 0x0A
+    BEEP                  = 0x0B
+    RFID_SEND             = 0x0C
+    RFID_SEND_COMPLETE    = 0x0D
+    RX_ERROR              = 0xFF
+
+class ResponseLength(Enum):
+    RESPONSE_ATR          = 5
+    TRANSMIT_OVERHEAD     = 3
+
+class Reader:
+
+    class ResetException(Exception):
+        """The reader has been unexpectedly reset."""
+        pass
+
+    class ReaderError(Exception):
+        """The reader did something wrong and MUST be reset."""
+        pass
+
+    class CorruptedPacketException(Exception):
+        pass
+
+    class Leds(Enum):
+        RED_LED     = 0
+        GREEN_LED   = 1
+        BLUE_LED    = 2
+
+    BAUDRATE   = 19200
+    BYTESIZE   = serial.EIGHTBITS
+    PARITY     = serial.PARITY_EVEN
+    STOPBITS   = serial.STOPBITS_ONE
+    RX_TIMEOUT = 0.5
+
+    MAX_BEEP_LENGTH = 8
+    MAX_RFID_PAYLOAD = 128
+
+    PacketHead = mystruct('PacketHead',
+                          ['id',     'length'],
+                          [ t.uint8,  t.uint8])
+
+    def __init__(self, port):
+        self.port = serial.Serial(port, self.BAUDRATE, self.BYTESIZE,
+                                  self.PARITY, self.STOPBITS, self.RX_TIMEOUT);
+
+    def _transmit_packet(self, packet_id, payload):
+        checksum =  packet_id.value
+        checksum ^= len(payload)
+        for byte in payload:
+            checksum ^= byte
+        self.port.write(bytes([packet_id.value, len(payload)]) + payload
+                        + bytes([checksum]))
+
+    def _checksum_mismatch(self, head, payload, checksum):
+        check =  head.id
+        check ^= head.length
+        for byte in payload:
+            check ^= byte
+        return check != checksum
+
+    def _expect_packet(self, length):
+        try:
+            p, payload = self.PacketHead.unpack_from(self.port.read(length))
+            payload, checksum = payload[:-1], payload[-1]
+            if self._checksum_mismatch(p, payload, checksum):
+                raise Reader.CorruptedPacketException()
+            return p, payload
+        except struct.error:
+            # The struct probably cannot be unpacked due to a corrupted
+            # packet
+            raise Reader.CorruptedPacketException()
+
+    def _retransmit_wrapper(self, packet_id, payload, resp_length):
+        """Transmits the packet, receives response while handling Rx errors.
```

Comment:
Preferably: """Transmits the packet & receives response, retransmitting if needed.""" (or something like that).